### PR TITLE
VTWAP subscriber and feed.q

### PR DESCRIPTION
**Pull request by SeanLukeMurphy**
- VTWAP subscriber added.
   - subscriber keeps all tables as well as a summary table.
   - summary table is keyed by sym and contains three fields time, price, size  
   - waps[] function takes arguments of a list of syms, start time and end time, and is used to generate the vwap and twap from the summary table.
- Additions to feed.q/database.q
   - New table added; srcquote
   - New function q2 which is used to generate rows for the new table srcquote
- Additions to start/stop script
   - Added the vtwap startup and the wap procname for kill.q

### CODE REVIEW

File: appconfig/passwords/accesslist.txt [hunk 1]
```diff
@@ -16,7 +16,5 @@ sortslave:pass
 metrics:pass
 vwapsub:pass
 checker:pass
-unittest:pass
-admin:admin
 vtwap:pass
-unittests:pass
```
Comment (ThomasSmyth) [inline appconfig/passwords/accesslist.txt]: Does this need added back in?

File: appconfig/process.csv [hunk 2]
```diff
@@ -20,5 +20,4 @@ localhost,{KDBBASEPORT}+17,sortslave,sortslave2
 localhost,{KDBBASEPORT}+18,metrics,metrics1
 localhost,{KDBBASEPORT}+19,iexfeed,iexfeed1
 localhost,{KDBBASEPORT}+20,checker,symcheck
-localhost,{KDBBASEPORT}+21,vtwap,vtwap1
-localhost,{KDBBASEPORT}+23,housekeeping,migrator1
```
Comment (ThomasSmyth) [inline appconfig/process.csv]: Does this need added back in?

File: code/processes/vtwap.q [hunk 3]
```diff
@@ -1,81 +1,64 @@
-tab:flip`time`sym`price`size`stop`cond`ex!();
+changetotab:{[t;x]
+  flip cols[t]!x                                                                     / Flip list into correct table schema
+  };
```
Comment (ThomasSmyth) [inline code/processes/vtwap.q]: 1 space indent instead of 2 or make this a one liner.

File: .gitignore [hunk 4]
```diff
@@ -1 +1,3 @@
+hdb/
 .DS_Store
```
Comment (SeanLukeMurphy) [inline .gitignore]: This did not ignore hdb as intended

File: code/processes/vtwap.q [hunk 5]
```diff
@@ -1,81 +1,64 @@
...
 
 upd:{[t;x]
-  if[t<>`trade;:()];
-  .vtwap.upd[t;tab upsert@[flip;x;enlist x]];
- };
+    .wap.tabfuncs[t][t;changetotab[t;x]];
+  };
```
Comment (ThomasSmyth) [inline code/processes/vtwap.q]: 1 space indent instead of 2 here + only use 2 spaces above.

File: code/processes/vtwap.q [hunk 6]
```diff
@@ -1,81 +1,64 @@
...
+subscribesyms:@[value;`subscribesyms;`];                                                        / A list of syms to subscribe for, (`) means all syms
+tpconnsleepintv:@[value;`tpconnsleepintv;10];                                                   / Number of seconds between attempts to connect to the tp
+summary:([sym:()]time:();price:();size:());                                                     / Summary table schema
+tabfuncs:(!/) flip 2 cut                                                                        / Define dictionary in convenient list format below
+    (
+    `trade;       {[t;x] @[`.wap.summary;asc([]sym:distinct x[`sym]);,';value exec (time;price;size) by sym from x];t insert x};
+    `quote;       {[t;x] t insert x};
+    `trade_iex;   {[t;x] @[`.wap.summary;asc([]sym:distinct x[`sym]);,';value exec (time;price;size) by sym from x];t insert x};
+    `quote_iex;   {[t;x] t insert x};
+		`srcquote;       {[t;x] t insert x}
```
Comment (ThomasSmyth) [inline code/processes/vtwap.q]: This line is out of alignment.

File: code/processes/vtwap.q [hunk 7]
```diff
@@ -1,81 +1,64 @@
...
+subscribeto:@[value;`subscribeto;`];	                                                          / A list of tables to subscribe to, default (`) means all tables
+subscribesyms:@[value;`subscribesyms;`];                                                        / A list of syms to subscribe for, (`) means all syms
+tpconnsleepintv:@[value;`tpconnsleepintv;10];                                                   / Number of seconds between attempts to connect to the tp
+summary:([sym:()]time:();price:();size:());                                                     / Summary table schema
+tabfuncs:(!/) flip 2 cut                                                                        / Define dictionary in convenient list format below
+    (
+    `trade;       {[t;x] @[`.wap.summary;asc([]sym:distinct x[`sym]);,';value exec (time;price;size) by sym from x];t insert x};
+    `quote;       {[t;x] t insert x};
+    `trade_iex;   {[t;x] @[`.wap.summary;asc([]sym:distinct x[`sym]);,';value exec (time;price;size) by sym from x];t insert x};
+    `quote_iex;   {[t;x] t insert x};
```
Comment (ThomasSmyth) [inline code/processes/vtwap.q]: Trades and quotes share funcs, could do:
```
tabfuncs:()!();
tabfuncs[`quote`quote_iex]:...
```

Should also avoid underscores in names.

File: code/processes/vtwap.q [hunk 8]
```diff
@@ -1,81 +1,64 @@
...
+    subinfo:.sub.subscribe[subscribeto;subscribesyms;schema;replaylog;first s];                 / Call subscribe function and save info
+    @[`.wap;key subinfo;:;value subinfo];                                                       / Setting subtables and tplogdate globals
     ];
- };
+	};
 
-notpconnected:{0=count select from .sub.SUBSCRIPTIONS where proctype in .vtwap.tickerplanttypes,active};
+upd:{[t;x]
+	tabfuncs[t][t;x];
+	}
```
Comment (ThomasSmyth) [inline code/processes/vtwap.q]: Indenting looks inconsistent compared to `subscribe` + needs ending semi colon.

File: code/processes/vtwap.q [hunk 9]
```diff
@@ -1,81 +1,64 @@
...
+
+notpconnected:{[]
+    :0=count select from .sub.SUBSCRIPTIONS where proctype in .wap.tickerplanttypes,active;			/ Check connection
+	};
```
Comment (ThomasSmyth) [inline code/processes/vtwap.q]: Odd indentation here, could also possibly be a one liner.

File: code/processes/vtwap.q [hunk 10]
```diff
@@ -1,81 +1,64 @@
...
 .servers.startup[];
-.vtwap.subscribe[];                                                                                    // subscribe to the tickerplant
-while[                                                                                                 // check if the tickerplant has connected, block the process until a connection is established
-  .vtwap.notpconnected[];
-  .os.sleep .vtwap.tpconnsleepintv;                                                                    // while no connected make the process sleep for X seconds and then run the subscribe function again
-  .servers.startup[];                                                                                  // run the servers startup code again (to make connection to discovery)
-  .vtwap.subscribe[];
+.wap.subscribe[];                                                                               / Subscribe to the tickerplant
+while[                                                                                        	/ Check if the tickerplant has connected, block the process until a connection is established
+	.wap.notpconnected[];
```
Comment (ThomasSmyth) [inline code/processes/vtwap.q]: Inconsistent.

File: code/processes/vtwap.q [hunk 11]
```diff
@@ -1,81 +1,64 @@
...
+  .os.sleep .wap.tpconnsleepintv;                                                               / While no connected make the process sleep for X seconds and then run the subscribe function again
+  .servers.startup[];                                                                         	/ Run the servers startup code again (to make connection to discovery)
```
Comment (ThomasSmyth) [inline code/processes/vtwap.q]: Comment alignment is off.

File: code/processes/vtwap.q [hunk 12]
```diff
@@ -1,81 +1,64 @@
...
   ];
 
-upd:.vtwap.upd;                                                                                        // set the upd function in the top level namespace
+upd:.wap.upd;
+
+waps:{[syms;st;et]																																							/ Calculate time/volume weighted average price			
+	syms:(),syms;
+	a:@'[x;i:{x[0]+til[x[1]-x[0]]} each {[x;st;et] x bin (st;et)}[;st;et] each x:.wap.summary'[syms;`time]];	
+	([]sym:syms;vwap:wavg'[@'[.wap.summary'[syms;`size];i];@'[.wap.summary'[syms;`price];i]];twap:wavg'[(next'[a]-a);@'[.wap.summary'[syms;`price];i]])
+	};
```
Comment (ThomasSmyth) [inline code/processes/vtwap.q]: Inconsistent alignment again.

File: code/processes/vtwap.q [hunk 13]
```diff
@@ -1,81 +1,64 @@
...
+  .servers.startup[];                                                                         	/ Run the servers startup code again (to make connection to discovery)
+  .wap.subscribe[];
   ];
 
-upd:.vtwap.upd;                                                                                        // set the upd function in the top level namespace
+upd:.wap.upd;
+
+waps:{[syms;st;et]																																							/ Calculate time/volume weighted average price			
+	syms:(),syms;
+	a:@'[x;i:{x[0]+til[x[1]-x[0]]} each {[x;st;et] x bin (st;et)}[;st;et] each x:.wap.summary'[syms;`time]];
```
Comment (ThomasSmyth) [inline code/processes/vtwap.q]: `x[1]-x[0]` can be `(-). x 1 0` OR replace the whol function with `{x+til y-x}./: ...`

When you have control over the params I think you should put the `each` last: `{[st;et;x] ...}[st;et] each ...`

File: code/processes/vtwap.q [hunk 14]
```diff
@@ -1,81 +1,64 @@
...
+  .wap.subscribe[];
   ];
 
-upd:.vtwap.upd;                                                                                        // set the upd function in the top level namespace
+upd:.wap.upd;
+
+waps:{[syms;st;et]																																							/ Calculate time/volume weighted average price			
+	syms:(),syms;
+	a:@'[x;i:{x[0]+til[x[1]-x[0]]} each {[x;st;et] x bin (st;et)}[;st;et] each x:.wap.summary'[syms;`time]];	
+	([]sym:syms;vwap:wavg'[@'[.wap.summary'[syms;`size];i];@'[.wap.summary'[syms;`price];i]];twap:wavg'[(next'[a]-a);@'[.wap.summary'[syms;`price];i]])
```
Comment (ThomasSmyth) [inline code/processes/vtwap.q]: This line is very long + confusing.

File: code/tick/feed.q [hunk 15]
```diff
@@ -65,13 +67,20 @@ q:{
  i:qx n:qn+til x;p:qp n;qn+:x;
  (s i;p-qb n;p+qa n;`long$vol x;`long$vol x;x?m;e i)}
 
-feed:{h$[rand 2;
- (".u.upd";`trade;t 1+rand maxn);
- (".u.upd";`quote;q 1+rand qpt*maxn)];}
+q2:{
+ if[not (qn+x)<count qx;batch len];
+ i:qx n:qn+til x;p:qp n;qn+:x;
+ (s i;count[i]?src;p-qb n;p+qa n;`long$vol x;`long$vol x;x?m;e i)}
```
Comment (ThomasSmyth) [inline code/tick/feed.q]: Should use minimum of 2 spaces to indent code + put closing brace on newline and add semi colon.

`` `long$vol x`` is repeated, could save as a var and only need to run it once.

File: code/tick/feed.q [hunk 16]
```diff
@@ -65,13 +67,20 @@ q:{
...
+
+feed:{$[rand 2;
+ h(".u.upd";`trade;t 1+rand maxn);
+ [h(".u.upd";`quote;q a:1+rand qpt*maxn);
+ h(".u.upd";`srcquote;q2 a:1+rand qpt*maxn)]];}
```
Comment (ThomasSmyth) [inline code/tick/feed.q]: More space indentation + closing brace stuff.

File: code/tick/feed.q [hunk 17]
```diff
@@ -65,13 +67,20 @@ q:{
...
 
-feedm:{h$[rand 2;
- (".u.upd";`trade;(enlist a#x),t a:1+rand maxn);
- (".u.upd";`quote;(enlist a#x),q a:1+rand qpt*maxn)];}
+feedm:{$[rand 2;
+ h(".u.upd";`trade;(enlist a#x),t a:1+rand maxn);
+ [h(".u.upd";`quote;(enlist a#x),q a:1+rand qpt*maxn);
+ h(".u.upd";`srcquote;(enlist a#x),q2 a:1+rand qpt*maxn)]];}
```
Comment (ThomasSmyth) [inline code/tick/feed.q]: Same spaces + braces.

File: appconfig/process.csv (hunk 2 above)
Comment (JamesBurrows) [reply appconfig/process.csv]: You'll need to keep `migrator1` in, at least. Otherwise the migrator script I assume we're keeping won't run through TorQ.

**Pull request closed without merging**